fvm/environment/uuids.go

```diff
@@ -39,45 +52,124 @@ func (generator ParseRestrictedUUIDGenerator) GenerateUUID() (uint64, error) {
+	// Since all blocks will have a transaction at txnIndex 0 but not
+	// necessarily a transaction at txnIndex 255, if we assign partition based
+	// only on txnIndex, partition 0's counter (and other low-valued
+	// partitions' counters) will fill up much more quickly than high-valued
+	// partitions' counters.  Therefore, a deterministically random offset is
+	// used to ensure the partitioned counters are roughly balanced.  Any byte
+	// in the sha hash is sufficiently random/uniform for this purpose (Note that
+	// block Id is already a sha hash, but its hash implementation may change
+	// underneath us).
+	//
+	// Note that since partition 0 reuses the legacy counter, its counter is
+	// much	further ahead than the other partitions.  If partition 0's counter
+	// is in danager of overflowing, use variants of "the power of two random
+	// choices" to shed load to other counters.
+	//
+	// The explicit mod is not really needed, but is there for completeness.
+	partitionOffset := sha256.Sum256(blockId[:])[0]
+	return byte((uint32(partitionOffset) + txnIndex) % 256)
 }
 
 func NewUUIDGenerator(
 	tracer tracing.TracerSpan,
+	log zerolog.Logger,
 	meter Meter,
 	txnState state.NestedTransactionPreparer,
+	blockHeader *flow.Header,
+	txnIndex uint32,
 ) *uUIDGenerator {
 	return &uUIDGenerator{
-		tracer:   tracer,
-		meter:    meter,
-		txnState: txnState,
+		tracer:      tracer,
+		log:         log,
+		meter:       meter,
+		txnState:    txnState,
+		blockHeader: blockHeader,
+		txnIndex:    txnIndex,
+		initialized: false,
 	}
 }
 
-// GetUUID reads uint64 byte value for uuid from the state
-func (generator *uUIDGenerator) getUUID() (uint64, error) {
-	stateBytes, err := generator.txnState.Get(flow.UUIDRegisterID)
+// getUint64 reads the uint64 value from the partitioned uuid register.
+func (generator *uUIDGenerator) getUint64() (uint64, error) {
+	stateBytes, err := generator.txnState.Get(generator.registerId)
 	if err != nil {
-		return 0, fmt.Errorf("cannot get uuid byte from state: %w", err)
+		return 0, fmt.Errorf(
+			"cannot get uuid partition %d byte from state: %w",
+			generator.partition,
+			err)
 	}
 	bytes := slices.EnsureByteSliceSize(stateBytes, 8)
 
 	return binary.BigEndian.Uint64(bytes), nil
 }
 
-// SetUUID sets a new uint64 byte value
-func (generator *uUIDGenerator) setUUID(uuid uint64) error {
+// setUint56 sets a new uint56 value into the partitioned uuid register.
+func (generator *uUIDGenerator) setUint56(
+	value uint64,
+) error {
+	if value > Uint56OverflowWarningThreshold {
+		if value > MaxUint56 {
+			return fmt.Errorf(
+				"uuid partition %d overflowed",
+				generator.partition)
+		}
+
+		generator.log.Warn().
+			Int("partition", int(generator.partition)).
+			Uint64("value", value).
+			Msg("uuid partition is running out of bits")
+	}
+
 	bytes := make([]byte, 8)
-	binary.BigEndian.PutUint64(bytes, uuid)
-	err := generator.txnState.Set(flow.UUIDRegisterID, bytes)
+	binary.BigEndian.PutUint64(bytes, value)
```

Comment:
I'm still clear here. Let's say before the change, the unpartitioned UUID has value `3` stored under the UUID key in storage. And now we use the partitioned UUID, tx1 get a counter value 0 from first UUID partition, and tx2 also get a counter value 0 from the second UUID partition. Since they get the same counter value, wouldn't cadence always get the same UUID value ? 

I think we also need to partition byte into the value as well. So that in the storage, it will look something like this:
```
uuid_partition_1_counter: 100000003
uuid_partition_4_counter: 400000005
```

This allows the uuid in different partition never have overlap, right?

Reply:
generator already returns with partition: https://github.com/onflow/flow-go/blob/e5e26dde0f4f5026dec5476a3a0bfaac7bef2d0b/fvm/environment/uuids.go#L197-L199

Reply:
Thanks for pointing that out. I makes sense now